### PR TITLE
refactor: isolate index handling to BulkDependentResource interface

### PR DESCRIPTION
**Pull request by metacosm**

### CODE REVIEW

**Comment by csviri** [comment]
pls revert this change on original branch

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/dependent/AbstractDependentResource.java [hunk 1]
```diff
@@ -20,15 +18,15 @@
...
+  private final boolean bulk = this instanceof BulkDependentResource;
 
   protected Creator<R, P> creator;
   protected Updater<R, P> updater;
-  protected BulkDependentResource<R, P> bulkDependentResource;
-
-  private final List<ResourceDiscriminator<R, P>> resourceDiscriminator = new ArrayList<>(1);
+  private final BulkDependentResource<R, P> bulkDependentResource;
+  private ResourceDiscriminator<R, P> resourceDiscriminator;
+  private int currentCount;
```
Comment (csviri) [inline operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/dependent/AbstractDependentResource.java:29]: Same problem here, what mentioned on other PR, that this cannot be a global variable

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/dependent/BulkDependentResource.java [hunk 2]
```diff
@@ -30,6 +32,21 @@
...
+   * @param primary the primary resource from which the desired state is inferred
+   * @param context the context in which the resource is being matched
+   * @return a {@link Result} encapsulating whether the resource matched its desired state and this
+   *         associated state if it was computed as part of the matching process. Use the static
+   *         convenience methods ({@link Result#nonComputed(boolean)} and
+   *         {@link Result#computed(boolean, Object)})
+   */
+  Result<R> match(R actualResource, P primary, int index, Context<P> context);
+
+  Optional<R> getSecondaryResource(P primary, int index, Context<P> context);
```
Comment (csviri) [inline operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/dependent/BulkDependentResource.java:50]: So maybe this should be `List<R> getSecondryResources(P primary, Context<P> context)`

**Comment by metacosm** [comment]
Superseded by #1517

**Pull request closed without merging**